### PR TITLE
added UserWarnings if max_epochs not set in the Trainer class

### LINKED ISSUE #10343: Default to infinite epochs, not 1000
Currently max_epochs defaults to 1000:

> If both max_epochs and max_steps aren't specified, max_epochs will default to 1000. To enable infinite training, set max_epochs = -1.

As a user, though, I would expect that if I don't specify a specific ending point, the training would continue indefinitely. In my own experiments, when the training cut off at 999 epochs, I was confused, and googling the issue didn't readily turn up this line in the documentation. When I checked my logs of all the hyperparams, max_epochs was set to None (I guess this override is applied internally). So as a user I feel like this is bad UX - I can't see a reason to put an arbitrary cutoff versus defaulting to infinite training.

### PR DESCRIPTION
**Pull request by Rajathbharadwaj**
## What does this PR do?
Intimates user if the max_epochs is not set in the trainer class

Fixes
#10343
## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/trainer/trainer.py [hunk 1]
```diff
@@ -444,6 +444,12 @@ def __init__(
         self.signal_connector = SignalConnector(self)
         self.tuner = Tuner(self)
 
+        # UserWarning for max_epochs if not set
+
+        warnings.warn("max_epochs not set, defaulted to 1000 epochs until stop.") if self.max_epochs is None else print(
+            f"max epochs set to {self.max_epochs}"
+        )
```
Comment (rohitgr7) [inline pytorch_lightning/trainer/trainer.py:451]:
```suggestion
        if self.max_epochs is None:
            warnings.warn("`max_epochs` is not set. By default it will run for 1000 epochs.")
        )
```

Comment (rohitgr7) [inline pytorch_lightning/trainer/trainer.py:451]:
```suggestion
        if self.max_epochs is None and max_steps == -1 and max_time is None:
            warnings.warn("`max_epochs` is not set. By default it will run for 1000 epochs.")
        )
```
print statement is not required.

File: docs/source/common/trainer.rst [hunk 2]
```diff
@@ -900,7 +900,7 @@ Stop training once this number of epochs is reached
     # default used by the Trainer
     trainer = Trainer(max_epochs=1000)
 
-If both ``max_epochs`` and ``max_steps`` aren't specified, ``max_epochs`` will default to ``1000``.
+If both ``max_epochs`` and ``max_steps`` aren't specified, ``max_epochs`` will default to ``1000`` and a ``UserWarning`` will be displayed stating the ``max_epochs`` is not set and defaulted to ``1000``.
```
Comment (rohitgr7) [inline docs/source/common/trainer.rst:903]:
```suggestion
If both ``max_epochs`` and ``max_steps`` aren't specified, ``max_epochs`` will default to ``1000``.
```
I don't think we need this since we already specified the behavior here.

**Comment by Rajathbharadwaj** [comment]
I've done the suggested changes @rohitgr7 thanks a ton for helping me out!

File: pytorch_lightning/trainer/trainer.py [hunk 3]
```diff
@@ -444,6 +444,11 @@ def __init__(
         self.signal_connector = SignalConnector(self)
         self.tuner = Tuner(self)
 
+        # UserWarning for max_epochs if not set
+
+        if self.max_epochs is None and max_steps == -1 and max_time is None:
+            warnings.warn("`max_epochs` is not set. By default it will run for 1000 epochs.")
```
Comment (tchaton) [inline pytorch_lightning/trainer/trainer.py:450]: Let's use rank_zero_warn :) And might be worth creating a utility to parse the max_epochs, max_steps before passing them to the Loops.
This would be cleaner.

Comment (Rajathbharadwaj) [reply pytorch_lightning/trainer/trainer.py]: @tchaton I have updated to the rank_zero_warn of type `UserWarning`. Also, if you can let me know what exact utility do we need to parse `max_epochs`, `max_steps` it'd be great, as I'm not sure I understood what you meant :")

Comment (tchaton) [reply pytorch_lightning/trainer/trainer.py]: Hey @Rajathbharadwaj,

I meant something like this

```py
# rank_zero_warning of type UserWarning for max_epochs if not set
if self.max_epochs is None and max_steps == -1 and max_time is None:
    rank_zero_warn(
        "`max_epochs` is not set. By default it will run for 1000 epochs.",
        UserWarning,
    )

fit_loop = FitLoop(
    min_epochs=(1 if (min_epochs is None and min_steps is None and max_time is None) else min_epochs),
    max_epochs=(
    max_epochs if max_epochs is not None else (1000 if (max_steps == -1 and max_time is None) else -1)
),
)
```

could become

```py
min_epochs, max_epochs = _parse_max_epochs_and_steps(max_steps, max_epochs, max_time, min_steps, min_epochs)

# utilities/...
def _parse_max_epochs_and_steps(max_steps, max_epochs, max_time, min_steps, min_epochs):
    if max_epochs is None and max_steps == -1 and max_time is None:
        rank_zero_warn(
            "`max_epochs` is not set. By default it will run for 1000 epochs.",
            UserWarning,
        )
    min_epochs = (1 if (min_epochs is None and min_steps is None and max_time is None) else min_epochs)
    max_epochs = max_epochs if max_epochs is not None else (1000 if (max_steps == -1 and max_time is None) else -1)
    return min_epochs, max_epochs
```

Comment (Rajathbharadwaj) [reply pytorch_lightning/trainer/trainer.py]: Oh, okay, gotcha! thanks for the help @tchaton

Comment (tchaton) [reply pytorch_lightning/trainer/trainer.py]: maybe pytorch_lightning/loops/utilities.py might be the most appropriate.

Comment (Rajathbharadwaj) [reply pytorch_lightning/trainer/trainer.py]: Thanks @tchaton 
I've added the necessary changes mentioned by you please checkout! 
Again, thanks for helping me out!

File: pytorch_lightning/trainer/trainer.py [hunk 4]
```diff
@@ -444,12 +445,12 @@ def __init__(
         self.signal_connector = SignalConnector(self)
         self.tuner = Tuner(self)
 
-        fit_loop = FitLoop(
-            min_epochs=(1 if (min_epochs is None and min_steps is None and max_time is None) else min_epochs),
-            max_epochs=(
-                max_epochs if max_epochs is not None else (1000 if (max_steps == -1 and max_time is None) else -1)
-            ),
+        # raises rank_zero_warn if max_epochs is not set
```
Comment (tchaton) [inline pytorch_lightning/trainer/trainer.py:448]:
```suggestion
```

**Review by tchaton — APPROVED** [review]
LGTM !

File: pytorch_lightning/loops/utilities.py [hunk 5]
```diff
@@ -62,6 +63,18 @@ def _extract_hiddens(training_step_output: STEP_OUTPUT, truncated_bptt_steps: in
     return hiddens
 
 
+# rank_zero_warning of type UserWarning for max_epochs if not set
+def _parse_max_epochs_and_steps(max_steps, max_epochs, max_time, min_steps, min_epochs):
```
Comment (tchaton) [inline pytorch_lightning/loops/utilities.py:67]: typing ?

Comment (Rajathbharadwaj) [reply pytorch_lightning/loops/utilities.py:67]: My bad, I will update. So used to not using typing 😄

File: pytorch_lightning/loops/utilities.py [hunk 6]
```diff
@@ -62,6 +64,20 @@ def _extract_hiddens(training_step_output: STEP_OUTPUT, truncated_bptt_steps: in
     return hiddens
 
 
+# rank_zero_warning of type UserWarning for max_epochs if not set
+def _parse_max_epochs_and_steps(
+    max_steps: int, max_epochs: int, max_time: Union[str, timedelta, Dict[str, int]], min_steps: int, min_epochs: int
+) -> Tuple[int, int]:
+    if max_epochs is None and max_steps == -1 and max_time is None:
+        rank_zero_warn(
+            "`max_epochs` is not set. By default it will run for 1000 epochs.",
```
Comment (awaelchli) [inline pytorch_lightning/loops/utilities.py:73]:
```suggestion
            "`max_epochs` is not set. By default it will run for 1000 epochs.",
```
The message is not accurate. 
1) "it" here refers to "max_epochs" from the previous sentence, but this is not what you meant to express.
2) The loop is not guaranteed to run for 1000 epochs.

```suggestion
            "`max_epochs` was not set. Setting it to 1000 epochs.",
```

File: pytorch_lightning/loops/utilities.py [hunk 7]
```diff
@@ -62,6 +64,20 @@ def _extract_hiddens(training_step_output: STEP_OUTPUT, truncated_bptt_steps: in
     return hiddens
 
 
+# rank_zero_warning of type UserWarning for max_epochs if not set
```
Comment (awaelchli) [inline pytorch_lightning/loops/utilities.py:67]: let's delete this comment :)
the code is self-documenting that already

File: pytorch_lightning/loops/utilities.py [hunk 8]
```diff
@@ -62,6 +64,20 @@ def _extract_hiddens(training_step_output: STEP_OUTPUT, truncated_bptt_steps: in
...
+def _parse_max_epochs_and_steps(
```
Comment (awaelchli) [inline pytorch_lightning/loops/utilities.py:68]:
```suggestion
def _parse_loop_limits(
```
I propose this name because this function does not only take max_epochs/steps as input.

File: pytorch_lightning/loops/utilities.py [hunk 9]
```diff
@@ -62,6 +64,20 @@ def _extract_hiddens(training_step_output: STEP_OUTPUT, truncated_bptt_steps: in
...
+    max_steps: int, max_epochs: int, max_time: Union[str, timedelta, Dict[str, int]], min_steps: int, min_epochs: int
+) -> Tuple[int, int]:
+    if max_epochs is None and max_steps == -1 and max_time is None:
```
Comment (awaelchli) [inline pytorch_lightning/loops/utilities.py:71]: We should rewrite it to this structure:

```
if condition:
    emit warning that max_epochs=1000
    set max_epochs=1000
```
This way it is clearer under which condition the warning applies and what the end value will be.

File: pytorch_lightning/loops/utilities.py [hunk 10]
```diff
@@ -62,6 +64,20 @@ def _extract_hiddens(training_step_output: STEP_OUTPUT, truncated_bptt_steps: in
     return hiddens
 
 
+# rank_zero_warning of type UserWarning for max_epochs if not set
+def _parse_max_epochs_and_steps(
+    max_steps: int, max_epochs: int, max_time: Union[str, timedelta, Dict[str, int]], min_steps: int, min_epochs: int
```
Comment (awaelchli) [inline pytorch_lightning/loops/utilities.py:69]: I would make the order with min* first, followed by max*, i.e.: 

min_steps, max_steps, min_epochs, max_epochs, max_time

**Review by awaelchli — CHANGES_REQUESTED** [review]
Thx for working on this!

File: pytorch_lightning/loops/utilities.py (hunk 9 above)
Comment (Rajathbharadwaj) [reply pytorch_lightning/loops/utilities.py:71]: If I understood correctly, then it should be the following

```
if max_epochs is None:
     rank_zero_warn(...)
     min_epochs = (1 if (min_epochs is None and min_steps is None and max_time is None) else min_epochs)
     max_epochs = max_epochs if max_epochs is not None else (1000 if (max_steps == -1 and max_time is None) 
        else -1)
    return min_epochs, max_epochs
```
Is this correct? Thanks

File: pytorch_lightning/loops/utilities.py (hunk 10 above)
Comment (Rajathbharadwaj) [reply pytorch_lightning/loops/utilities.py:69]: Gotcha!

**Comment by Rajathbharadwaj** [comment]
Hey @awaelchli, I've updated the files with the necessary changes, please check it out and let me know, and also on the above mention doubt. Thanks!

File: pytorch_lightning/trainer/trainer.py [hunk 11]
```diff
@@ -445,8 +445,8 @@ def __init__(
         self.signal_connector = SignalConnector(self)
         self.tuner = Tuner(self)
 
-        min_epochs, max_epochs = _parse_max_epochs_and_steps(
-            self.max_steps, self.max_epochs, max_time, self.min_steps, self.min_epochs
+        min_epochs, max_epochs = _parse_loop_limits(
+            self.min_steps, self.max_steps, self.min_epochs, self.max_epochs, max_time
```
Comment (Rajathbharadwaj) [inline pytorch_lightning/trainer/trainer.py:449]: I'm sorry, I forgot to update the call here 😄

**Comment by Rajathbharadwaj** [comment]
Hey @tchaton is there anything left to do for this PR?

File: pytorch_lightning/loops/utilities.py [hunk 12]
```diff
@@ -62,6 +64,41 @@ def _extract_hiddens(training_step_output: STEP_OUTPUT, truncated_bptt_steps: in
...
+    Returns:
+        A list of tuples (opt_idx, optimizer) of currently active optimizers.
+    """
+    if max_epochs is None:
+        if max_steps == -1 and max_time is None:
+            rank_zero_warn(
+                "`max_epochs` was not set. Setting it to 1000 epochs.",
+                UserWarning,
+            )
+            max_epochs = default_max_epochs
```
Comment (rohitgr7) [inline pytorch_lightning/loops/utilities.py:95]:
```suggestion
) -> Tuple[Optional[int], int]:
    """This utility computes the number minimum and maximum number of epochs " based on the minimum and maximum number
    of steps and maximum time.

    Args:
        min_steps: Minimum number of steps
        max_steps: Maximum number of steps
        min_epochs: Minimum number of epochs
        max_epochs: Maximum number of epochs
        max_time: Maximum time for the training.

    Returns:
        A list of tuples (opt_idx, optimizer) of currently active optimizers.
    """
    if max_epochs is None:
        if max_steps == -1 and max_time is None:
            rank_zero_warn(
                "`max_epochs` was not set. Setting it to 1000 epochs.",
                UserWarning,
            )
            max_epochs = 1000
```
no need for the default argument.

**Review by rohitgr7 — APPROVED** [review]

**Review by awaelchli — APPROVED** [review]
